deepspeed/pt/deepspeed_cpu_adam.py

```diff
@@ -0,0 +1,200 @@
+import math
+import torch
+
+
+class CPUAdam(torch.optim.Optimizer):

```

Comment:
Is it possible to implement this in a way that does not require reimplementing the optimizer? I know that it does not matter much at the moment since ZeRO is only implemented for Adam, but I think it's worthwhile future proofing if feasible. And avoiding code duplication is would be ideal.

To my understanding, the changes that this makes to Adam are concerned with getting flattened buffers on the CPU from ZeRO. Could we instead do this work on the ZeRO side so that `self.param_groups` and `self.state` appear normal to the optimizer? Then I think Adam could work unmodified? I'm not very close to this part of DeepSpeed; please tell me if this is nonsense :-).

Knowing that not every optimizer may play nicely and someone may want to modify something specifically for an optimizer, it would still be wonderful to avoid the reimplementation issue. I wonder if it is possible to implement this as a subclass and use `super()` to get extensibility:

```python
# base class
class CPUOptimizer(torch.optim.Optimizer):
    def step(data_from_zero):
        # sneaky replacement of CPU tensors
        cpu_stuff(data_from_zero)

        # now do the real step() work
        super().step()

# We can provide deepspeed.CPUAdam at first for convenience
# step() will first call the CPUOptimizer step() and then Adam step()"""
class CPUAdam(deepspeed.CPUOptimizer, torch.optim.Adam):
    pass

# potential client code:
class CPUFancyOptimizer(deepspeed.CPUOptimizer, FancyOptimizer):
    def step(data_from_zero):
        custom_copy_stuff()

        # do the rest. This will call CPUOptimizer step() and then FancyOptimizer's step()
        super().step(data_from_zero)
```

Reply:
Thanks for this feedback. We now use torch.optim.Adam directly.